I ran into issues with that as well because it's not possible to implement the Layer trait on a foreign type

```
impl<S, A, B> tower::Layer<S> for Either<A, B>
where
	A: tower::Layer<S>,
	B: tower::Layer<S>,
{
	type Service = Either<A::Service, B::Service>;

	fn layer(&self, inner: S) -> Self::Service {
		match self {
			Either::Left(layer) => Either::A(layer.layer(inner)),
			Either::Right(layer) => Either::B(layer.layer(inner)),
		}
	}
}
```

```bash
error[E0210]: type parameter `S` must be used as the type parameter for some local type (e.g., `MyStruct<S>`)
   --> server/src/middleware/rpc/mod.rs:142:6
    |
142 | impl<S, A, B> tower::Layer<S> for Either<A, B>
    |      ^ type parameter `S` must be used as the type parameter for some local type
    |
    = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local
    = note: only traits defined in the current crate can be implemented for a type parameter

For more information about this error, try `rustc --explain E0210`.
warning: `jsonrpsee-server` (lib) generated 5 warnings
error: could not compile `jsonrpsee-server` (lib) due to previous error; 5 warnings emitted


```

--- a/server/Cargo.toml
+++ b/server/Cargo.toml
@@ -30,6 +30,7 @@ tower = { version = "0.4.13", features = ["util"] }
 route-recognizer = "0.3.1"
 http = "0.2.9"
 thiserror = "1.0.44"
+pin-project = "1.1.3"
 
 [dev-dependencies]
 anyhow = "1"

--- a/server/src/middleware/rpc/layers.rs
+++ b/server/src/middleware/rpc/layers.rs
@@ -24,8 +24,11 @@
 // IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
+use std::pin::Pin;
 use std::sync::Arc;
+use std::task::{Context, Poll};
 
+use futures_util::{ready, Future};
 use jsonrpsee_core::server::{
 	BoundedSubscriptions, MethodCallback, MethodResponse, MethodSink, Methods, SubscriptionState,
 };
@@ -33,6 +36,8 @@ use jsonrpsee_core::tracing::{rx_log_from_json, tx_log_from_str};
 use jsonrpsee_core::traits::IdProvider;
 use jsonrpsee_types::error::{reject_too_many_subscriptions, ErrorCode};
 use jsonrpsee_types::{ErrorObject, Request};
+use pin_project::pin_project;
+use tower::BoxError;
 
 use super::{RpcServiceT, TransportProtocol};
 
@@ -170,5 +175,66 @@ where
 		let rp = self.service.call(request, transport).await;
 		tx_log_from_str(&rp.result, self.max);
 		rp
+	}
+}
+
+/// Similar to [`tower::util::Either`] but
+/// adjusted to satisfy the trait bound [`RpcServiceT].
+//
+// NOTE: This is introduced because it doesn't
+// work to implement tower::Layer for
+// external types such as future::Either.
+#[pin_project(project = EitherProj)]
+#[derive(Clone, Debug)]
+pub enum Either<A, B> {
+	/// One type of backing [`RpcServiceT`].
+	A(#[pin] A),
+	/// The other type of backing [`RpcServiceT`].
+	B(#[pin] B),
+}
+
+impl<A, B, T, AE, BE> Future for Either<A, B>
+where
+	A: Future<Output = Result<T, AE>>,
+	AE: Into<BoxError>,
+	B: Future<Output = Result<T, BE>>,
+	BE: Into<BoxError>,
+{
+	type Output = Result<T, BoxError>;
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		match self.project() {
+			EitherProj::A(fut) => Poll::Ready(Ok(ready!(fut.poll(cx)).map_err(Into::into)?)),
+			EitherProj::B(fut) => Poll::Ready(Ok(ready!(fut.poll(cx)).map_err(Into::into)?)),
+		}
+	}
+}
+
+impl<S, A, B> tower::Layer<S> for Either<A, B>
+where
+	A: tower::Layer<S>,
+	B: tower::Layer<S>,
+{
+	type Service = Either<A::Service, B::Service>;
+
+	fn layer(&self, inner: S) -> Self::Service {
+		match self {
+			Either::A(layer) => Either::A(layer.layer(inner)),
+			Either::B(layer) => Either::B(layer.layer(inner)),
+		}
+	}
+}
+
+#[async_trait::async_trait]
+impl<'a, A, B> RpcServiceT<'a> for Either<A, B>
+where
+	A: RpcServiceT<'a> + Send + Sync,
+	B: RpcServiceT<'a> + Send + Sync,
+{
+	async fn call(&self, request: Request<'a>, transport: TransportProtocol) -> MethodResponse {
+		match self {
+			Either::A(service) => service.call(request, transport).await,
+			Either::B(service) => service.call(request, transport).await,
+		}
 	}
 }

--- a/server/src/middleware/rpc/mod.rs
+++ b/server/src/middleware/rpc/mod.rs
@@ -28,7 +28,6 @@
 
 mod layers;
 
-use futures_util::future::Either;
 pub use layers::*;
 
 use tower::layer::util::{Identity, Stack};
@@ -90,7 +89,7 @@ impl<L> RpcServiceBuilder<L> {
 	///
 	/// See the documentation for [`tower::ServiceBuilder::option_layer`] for more details.
 	pub fn option_layer<T>(self, layer: Option<T>) -> RpcServiceBuilder<Stack<Either<T, Identity>, L>> {
-		let layer = if let Some(layer) = layer { Either::Left(layer) } else { Either::Right(Identity::new()) };
+		let layer = if let Some(layer) = layer { Either::A(layer) } else { Either::B(Identity::new()) };
 		self.layer(layer)
 	}
 
@@ -122,19 +121,5 @@ impl<L> RpcServiceBuilder<L> {
 		L: tower::Layer<S>,
 	{
 		self.0.service(service)
-	}
-}
-
-#[async_trait::async_trait]
-impl<'a, A, B> RpcServiceT<'a> for Either<A, B>
-where
-	A: RpcServiceT<'a> + Send + Sync,
-	B: RpcServiceT<'a> + Send + Sync,
-{
-	async fn call(&self, request: Request<'a>, transport: TransportProtocol) -> MethodResponse {
-		match self {
-			Either::Left(service) => service.call(request, transport).await,
-			Either::Right(service) => service.call(request, transport).await,
-		}
 	}
 }

